src/types/mutex.rs

```diff
@@ -0,0 +1,86 @@
+use std::ops::Deref;
+
+/// Wrapper for [`PyMutex`](https://docs.python.org/3/c-api/init.html#c.PyMutex) exposing an RAII interface.
+#[derive(Debug)]
+pub struct PyMutex<T> {
+    _mutex: crate::ffi::PyMutex,
+    data: T,
+}
+
+/// RAII guard to handle releasing a PyMutex lock.
+#[derive(Debug)]
+pub struct PyMutexGuard<'a, T> {
+    _mutex: &'a mut crate::ffi::PyMutex,
+    data: &'a T,
```

Comment:
```suggestion
    mutex: &'a PyMutex<T>,
```

... and then can read from inside the mutex's `UnsafeCell` for the `Deref` and `DerefMut` implementations.